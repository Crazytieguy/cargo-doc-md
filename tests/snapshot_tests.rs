use cargo_doc_md::{converter, parser, ConversionOptions};
use std::path::Path;

#[test]
fn test_multifile_output_structure() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let mut file_paths: Vec<_> = output.files.keys().collect();
    file_paths.sort();

    insta::assert_debug_snapshot!("multifile_structure", file_paths);
    assert_eq!(output.crate_name, "test_crate");
}

#[test]
fn test_index_file() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let index_content = output.files.get("index.md").expect("index.md not found");
    insta::assert_snapshot!("index_md", index_content);
}

#[test]
fn test_lib_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    if let Some(lib_content) = output.files.get("test_crate.md") {
        insta::assert_snapshot!("lib_module", lib_content);
    }
}

#[test]
fn test_types_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    if let Some(types_content) = output.files.get("types.md") {
        insta::assert_snapshot!("types_module", types_content);
    }
}

#[test]
fn test_functions_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    if let Some(functions_content) = output.files.get("functions.md") {
        insta::assert_snapshot!("functions_module", functions_content);
    }
}

#[test]
fn test_nested_modules() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    if let Some(nested_content) = output.files.get("nested.md") {
        insta::assert_snapshot!("nested_module", nested_content);
    }

    if let Some(inner_content) = output.files.get("nested/inner.md") {
        insta::assert_snapshot!("nested_inner_module", inner_content);
    }

    if let Some(deep_content) = output.files.get("nested/inner/deep.md") {
        insta::assert_snapshot!("nested_deep_module", deep_content);
    }
}

#[test]
fn test_struct_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("## test_crate::UnitStruct"));
    assert!(lib_content.contains("**Unit Struct**"));

    assert!(lib_content.contains("## test_crate::TupleStruct"));
    assert!(lib_content.contains("**Tuple Struct**"));

    assert!(lib_content.contains("## test_crate::PlainStruct"));
    assert!(lib_content.contains("**Fields:**"));
}

#[test]
fn test_generic_struct_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("## test_crate::GenericStruct"));
    assert!(lib_content.contains("**Generic Parameters:**"));
}

#[test]
fn test_enum_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("## test_crate::SimpleEnum"));
    assert!(lib_content.contains("**Variants:**"));

    assert!(lib_content.contains("## test_crate::ComplexEnum"));
}

#[test]
fn test_function_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let functions_content = output.files.get("functions.md").expect("functions module not found");

    assert!(functions_content.contains("fn add"));
    assert!(functions_content.contains("fn multiply"));
}

#[test]
fn test_trait_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("## test_crate::MyTrait"));
    assert!(lib_content.contains("**Type:** Trait"));
}

#[test]
fn test_methods_and_impls() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("**Methods:**"));
}

#[test]
fn test_constants_and_type_aliases() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("MAX_SIZE"));
    assert!(lib_content.contains("Result"));
}

#[test]
fn test_include_private() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");

    let output_public = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");
    let output_private = converter::convert_to_markdown_multifile(&crate_data, true)
        .expect("Failed to convert to markdown");

    let public_count = output_public.files.len();
    let private_count = output_private.files.len();

    assert!(private_count >= public_count,
        "Private output should have at least as many files as public output");
}

#[test]
fn test_conversion_end_to_end() {
    let output_dir = std::env::temp_dir().join("cargo_doc_md_test");
    let _ = std::fs::remove_dir_all(&output_dir);

    let json_path = Path::new("tests/fixtures/test_crate.json");
    let options = ConversionOptions {
        input_path: json_path,
        output_dir: &output_dir,
        include_private: false,
    };

    cargo_doc_md::convert_json_file(&options).expect("Conversion failed");

    assert!(output_dir.join("test_crate").exists());
    assert!(output_dir.join("test_crate/index.md").exists());

    std::fs::remove_dir_all(&output_dir).ok();
}

#[test]
fn test_legacy_single_file_output() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let markdown = converter::convert_to_markdown(&crate_data, false)
        .expect("Failed to convert to markdown");

    assert!(markdown.contains("# test_crate"));
    assert!(markdown.contains("## Table of Contents"));
    insta::assert_snapshot!("legacy_single_file", markdown);
}
