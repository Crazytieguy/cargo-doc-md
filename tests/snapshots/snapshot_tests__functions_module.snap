---
source: tests/snapshot_tests.rs
expression: functions_content
---
**test_crate > functions**

# Module: functions

## Contents

**Functions**

- [`add`](#add) - Adds two numbers together.
- [`async_function`](#async_function) - An async function that simulates fetching data.
- [`complex_generics`](#complex_generics) - A function that takes multiple generic parameters with different bounds.
- [`const_function`](#const_function) - A const function that can be evaluated at compile time.
- [`filter`](#filter) - Filters a slice based on a predicate.
- [`for_each`](#for_each) - Applies a closure to each element in a slice.
- [`higher_order_function`](#higher_order_function) - A higher-order function that applies a function to a value.
- [`map`](#map) - Maps a slice to a new vector using a closure.
- [`multiply`](#multiply) - Multiplies two values that implement `Mul`.
- [`process_mut_slice`](#process_mut_slice) - Mutates a byte slice in place.
- [`process_slice`](#process_slice) - Processes a byte slice and returns a new vector.
- [`unsafe_function`](#unsafe_function) - An unsafe function that dereferences a raw pointer.

---

## test_crate::functions::add

*Function*

Adds two numbers together.

# Examples

```
use test_crate::functions::add;

assert_eq!(add(2, 3), 5);
```

```rust
fn add(a: i32, b: i32) -> i32
```



## test_crate::functions::async_function

*Function*

An async function that simulates fetching data.

# Arguments

* `url` - The URL to fetch from

# Returns

A `Result` containing the fetched string or an error message.

# Examples

```no_run
# async fn example() {
use test_crate::functions::async_function;

let result = async_function("https://example.com").await;
assert!(result.is_ok());
# }
```

```rust
fn async_function(url: &str) -> Result<String, String>
```



## test_crate::functions::complex_generics

*Function*

A function that takes multiple generic parameters with different bounds.

```rust
fn complex_generics<T, U, V>(t: T, u: U, _v: V) -> String
```



## test_crate::functions::const_function

*Function*

A const function that can be evaluated at compile time.

# Examples

```
use test_crate::functions::const_function;

const VALUE: i32 = const_function(21);
assert_eq!(VALUE, 42);
```

```rust
fn const_function(x: i32) -> i32
```



## test_crate::functions::filter

*Function*

Filters a slice based on a predicate.

```rust
fn filter<T, F>(slice: &[T], predicate: F) -> Vec<&T>
```



## test_crate::functions::for_each

*Function*

Applies a closure to each element in a slice.

# Type Parameters

* `T` - The type of elements in the slice
* `F` - The closure type

```rust
fn for_each<T, F>(slice: &[T], f: F)
```



## test_crate::functions::higher_order_function

*Function*

A higher-order function that applies a function to a value.

# Arguments

* `f` - A function that takes an `i32` and returns an `i32`

# Returns

The result of calling `f(42)`.

```rust
fn higher_order_function<F>(f: F) -> i32
```



## test_crate::functions::map

*Function*

Maps a slice to a new vector using a closure.

```rust
fn map<T, U, F>(slice: &[T], f: F) -> Vec<U>
```



## test_crate::functions::multiply

*Function*

Multiplies two values that implement `Mul`.

This is a generic function that works with any type implementing
the multiplication operator.

# Type Parameters

* `T` - A type that implements `Mul` and `Copy`

```rust
fn multiply<T>(a: T, b: T) -> T
```



## test_crate::functions::process_mut_slice

*Function*

Mutates a byte slice in place.

Each byte is incremented by 1 (with wrapping).

# Arguments

* `data` - A mutable reference to the byte slice

```rust
fn process_mut_slice(data: & mut [u8])
```



## test_crate::functions::process_slice

*Function*

Processes a byte slice and returns a new vector.

# Arguments

* `data` - The input byte slice to process

# Returns

A new `Vec<u8>` containing a copy of the input data.

```rust
fn process_slice(data: &[u8]) -> Vec<u8>
```



## test_crate::functions::unsafe_function

*Function*

An unsafe function that dereferences a raw pointer.

# Safety

The caller must ensure that `ptr` is valid and properly aligned.

# Arguments

* `ptr` - A raw pointer to a `u8`

```rust
fn unsafe_function(ptr: *const u8) -> u8
```
